src/azul/terra.py

```diff
@@ -487,19 +497,21 @@ def run_sql(self, query: str) -> BigQueryRows:
             result = job.result()
         else:
             delays = (10, 20, 40, 80)
+            attempts = len(delays) + 1
             assert sum(delays) < config.contribution_lambda_timeout(retry=False)
             for attempt, delay in enumerate((*delays, None)):
                 job: QueryJob = bigquery.query(query)
                 try:
                     result = job.result()
                 except Forbidden as e:
-                    if 'Exceeded rate limits' in e.message and delay is not None:
-                        log.warning('Exceeded BigQuery rate limit during attempt %i/%i. '
-                                    'Retrying in %is.',
-                                    attempt + 1, len(delays) + 1, delay, exc_info=e)
-                        sleep(delay)
-                    else:
-                        raise e
+                    handle_exception(e=e,
```

Comment:
Still too much duplicated code. Local nested function have access to local variables of the containing function.

Reply:
I don't like the `match` variable thingy, either. 

Try restructuring the control flow so that you don't need a nested function.